File: C_Projects/Asignacion #3 (Comparacion de Lenguajes)/linked_list.c
/*
 * linked_list.c
 * By Antonio F. Huertas
 * Edited by Saulo Almeda Nieves
 * 841-13-0170
 * 11 / 2 / 2015
 * This program defines and manipulates a linked list.
 */

#include <stdio.h>   // for puts, printf, scanf
#include <stdbool.h> // for bool, true, false;
#include <stddef.h>  // for NULL
#include <stdlib.h>  // for malloc, free

// Defines an alias for the node structure.
typedef struct node Node;

// Declares a structure for nodes.
struct node {
    int data;
    Node *next;
};

// Defines the data structure for a list.
typedef struct {
    Node *head;
} List;

// Declares functions using prototypes.
List *create_list(void);
void add_list(List *lst, int val);
void show_list(const List *lst);
int sum_list(const List *lst);
bool search_list(const List *lst, int val);
void destroy_list(List *lst);

//Added function prototypes
int length_list(const List *lst);
int minimum_list(const List *lst);

// Starts the execution of the program.
int main(void) {
    List *lst = create_list();
    puts("A linked list has been created.");

    add_list(lst, 40);
    add_list(lst, 10);
    add_list(lst, 20);
    add_list(lst, 50);
    add_list(lst, 30);

    printf("After adding some elements, the linked list is ");
    show_list(lst);
    printf("\nThe sum of its elements is %d\n\n", sum_list(lst));

    int num;
    printf("Enter a number to be searched: ");
    scanf("%d", &num);

    bool found = search_list(lst, num);
    if (found)
        puts("The number was found!");
    else
        puts("The number was NOT found.");

    printf("\nEnter a number to be added to the list: ");
    scanf("%d", &num);
    add_list(lst, num);

    printf("After adding a new element, the linked list is ");
    show_list(lst);
    printf("\nThe sum of its elements is %d\n\n", sum_list(lst));

    printf("\nThe length of the linked list is: %d\n\n", length_list(lst));

    printf("\nThe minimum element of the linked list is: %d\n\n", minimum_list(lst));

    destroy_list(lst);
    lst = NULL;
    puts("The linked list has been deallocated.");
    return 0;
}

//Returns the number of elements in the given list.
int length_list(const List *lst) {
    int len = 1;
    Node *curr = lst->head;
    while(curr->next != NULL) {
        curr = curr->next;
        len++;
    }
    return len;
}

//Returns the minimum element in the given list.
int minimum_list(const List *lst) {
    int min = lst->head->data;
    Node *curr = lst->head->next;
    while(curr->data < min) {
        min = curr->data;
        curr = curr->next;
    }
    return min;
}

// Returns a new empty list.
List *create_list(void) {
    List *lst = malloc(sizeof(List));
    lst->head = NULL;
    return lst;
}

// Adds a new value to the given list.
void add_list(List *lst, int val) {
    Node *temp = malloc(sizeof(Node));
    temp->data = val;
    temp->next = lst->head;
    lst->head = temp;
}

// Shows the contents of the given list.
void show_list(const List *lst) {
    for (Node *curr = lst->head; curr != NULL; curr = curr->next)
        printf("%d ", curr->data);
}

// Returns the sum of the given list.
int sum_list(const List *lst) {
    int sum = 0;
    for (Node *curr = lst->head; curr != NULL; curr = curr->next)
        sum += curr->data;
    return sum;
}

// Determines whether the given list contains the given value.
bool search_list(const List *lst, int val) {
    for (Node *curr = lst->head; curr != NULL; curr = curr->next)
        if (val == curr->data)
            return true;
    return false;
}

// Frees the memory allocated for the given list.
void destroy_list(List *lst) {
    while (lst->head != NULL) {
        Node *curr = lst->head;
        lst->head = lst->head->next;
        free(curr);
    }
    free(lst);
}

File: F# Workspace/Asignacion 5 (Comparacion de Lenguajes)/Asignacion 5 (Comparacion de Lenguajes)/linked_list.fs
(*
 * linked_list.fs
 * By Antonio F. Huertas
 * Edited by Saulo Almeda Nieves
 * 841-13-0170
 * 11/28/2015
 * This program defines and manipulates a linked list of integers.
 *)

// Defines a discriminated union type for a linked list.   
type IntList =
    | Nil
    | Cons of int * IntList

// sum_list : IntList -> int
// Returns the sum of the elements in the given list.
let rec sum_list = function
    | Nil -> 0
    | Cons (hd, tl) -> hd + sum_list tl

// contains_list : IntList -> int -> bool
// Returns true if the given list contains the given element.
let rec contains_list lst elem =
    match lst with
    | Nil -> false
    | Cons (hd, _) when elem = hd -> true
    | Cons (hd, tl) -> contains_list tl elem

// show_list : IntList -> unit
// Shows the elements in the given list.
let rec show_list = function
    | Nil -> ()
    | Cons (hd, tl) -> printf "%d " hd
                       show_list tl

// length_list : IntList -> int
// Returns the number of elements in the given list.
let rec length_list = function
    | Nil -> 0
    | Cons (hd, tl) -> 1 + length_list tl

// minimum_list : IntList -> int
// Returns the minimum element of the given list.
let rec minimum_list = function
    | Nil -> failwith "Error, empty list"
    | Cons (hd, Nil) -> hd
    | Cons (hd, tl) when hd < (minimum_list tl) -> hd
    | Cons (hd, tl) -> (minimum_list tl)

// append_list : IntList -> IntList -> IntList
// Returns the concatenation of the given lists.
let rec append_list lst lst2 =
    match lst with
    | Nil -> lst2
    | Cons (hd, Nil) -> Cons (hd, lst2)
    | Cons (hd, tl) -> Cons (hd, append_list tl lst2)

// reverse_list : IntList -> IntList
// Returns the given list in reversed order.
let rec reverse_list = function
    | Nil -> Nil
    | Cons (hd, Nil) -> Cons (hd, Nil)
    | Cons (hd, tl) -> append_list (reverse_list tl) (Cons (hd, Nil))

// remove_list : -> IntList -> int -> IntList
// Returns a list with a removed element.
let rec remove_list lst elem =
    match lst with
    | Nil -> failwith "Error, empty list."
    | Cons (hd, tl) when hd = elem -> tl
    | Cons (hd, tl) -> (append_list (Cons (hd, Nil)) (remove_list tl elem))

// start : unit -> unit
let start () =
    let lst = Cons (20, Cons (50, Cons (10, Cons (40, Cons (30, Nil)))))

    printf "This is the list: "
    show_list lst
    printfn "\nThis is the list, internally: %A" lst
    printfn "The sum of its elements is %d" (sum_list lst)
    printfn "Does it contain 50? %b" (contains_list lst 50)
    printfn "The length of the list is: %d" (length_list lst)
    printfn "The minimum of the list is: %d" (minimum_list lst)
    printfn "The reversed list: %A" (reverse_list lst)

    printfn "Creating a second list..."
    let lst2 = Cons (23, Cons (50, Cons (11, Cons (41, Cons (30, Nil)))))
    printf "This is the list #2: "
    show_list lst2
    printfn "\n"
    printf "This are the two lists concatenated together: "
    (show_list (append_list lst lst2))
    printf "\nThis is the concatenated list without the number 50 is: " 
    show_list (remove_list (append_list lst lst2) 50)
    printf "\n"